In [2]:
import pandas as pd 

## 1. Data Gathering

In [ ]:
data = pd.read_csv('data/huge_1M_titanic.csv')

## 2. Data Preprocessing

In [73]:
data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1310,1,1,"Name1310, Miss. Surname1310",female,NaN,0,0,SOTON/O2 3101272,76.760165,NaN,C
1,1311,0,3,"Name1311, Col. Surname1311",male,29.0,0,0,223596,10.193097,NaN,S
2,1312,0,3,"Name1312, Mr. Surname1312",male,20.0,0,0,54636,12.029416,C83,C
3,1313,0,3,"Name1313, Mr. Surname1313",male,27.0,0,0,PC 17760,13.429448,NaN,S
4,1314,0,3,"Name1314, Mr. Surname1314",male,32.0,0,0,364512,4.840769,E33,C


In [74]:
data.shape

(1000000, 12)

In [75]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 12 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   PassengerId  1000000 non-null  int64  
 1   Survived     1000000 non-null  int64  
 2   Pclass       1000000 non-null  int64  
 3   Name         1000000 non-null  object 
 4   Sex          1000000 non-null  object 
 5   Age          801400 non-null   float64
 6   SibSp        1000000 non-null  int64  
 7   Parch        1000000 non-null  int64  
 8   Ticket       1000000 non-null  object 
 9   Fare         1000000 non-null  float64
 10  Cabin        229805 non-null   object 
 11  Embarked     997760 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 91.6+ MB


In [76]:
data.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,1.000000e+06,1000000.000000,1000000.00000,801400.000000,1000000.000000,1000000.000000,1000000.000000
mean,5.013095e+05,0.381681,2.31125,29.280339,0.511101,0.360037,33.461097
std,2.886753e+05,0.485799,0.83577,12.744978,1.104846,0.767296,47.976641
min,1.310000e+03,0.000000,1.00000,0.000000,0.000000,0.000000,0.000000
25%,2.513098e+05,0.000000,2.00000,21.000000,0.000000,0.000000,8.568924
50%,5.013095e+05,0.000000,3.00000,29.000000,0.000000,0.000000,15.883647
75%,7.513092e+05,1.000000,3.00000,37.000000,1.000000,0.000000,35.561000
max,1.001309e+06,1.000000,3.00000,114.000000,8.000000,6.000000,640.883614


In [7]:
data.isnull().sum()

PassengerId         0
Survived            0
Pclass              0
Name                0
Sex                 0
Age            198600
SibSp               0
Parch               0
Ticket              0
Fare                0
Cabin          770195
Embarked         2240
dtype: int64

In [8]:
data = data.drop(columns=['PassengerId','Name','Age','Ticket','Cabin'])

In [9]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked
0,1,1,female,0,0,76.760165,C
1,0,3,male,0,0,10.193097,S
2,0,3,male,0,0,12.029416,C
3,0,3,male,0,0,13.429448,S
4,0,3,male,0,0,4.840769,C


In [10]:
data.isnull().sum()

Survived       0
Pclass         0
Sex            0
SibSp          0
Parch          0
Fare           0
Embarked    2240
dtype: int64

In [11]:
data['Embarked'].mode()

0    S
Name: Embarked, dtype: object

#### Embarked : 
1. S - Southampton, England
2. C - Cherbourg, France
3. Q - Queenstown, Ireland

In [12]:
data['Embarked'].value_counts()

Embarked
S    729468
C    186846
Q     81446
Name: count, dtype: int64

## 3. Data Cleaning

In [13]:
data['Embarked'] = data['Embarked'].replace({'S': 'Southampton','C': 'Cherbourg', 'Q': 'Queenstown'})

In [14]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked
0,1,1,female,0,0,76.760165,Cherbourg
1,0,3,male,0,0,10.193097,Southampton
2,0,3,male,0,0,12.029416,Cherbourg
3,0,3,male,0,0,13.429448,Southampton
4,0,3,male,0,0,4.840769,Cherbourg


In [15]:
data['Embarked'].value_counts()

Embarked
Southampton    729468
Cherbourg      186846
Queenstown      81446
Name: count, dtype: int64

In [16]:
data.dropna(subset=['Embarked'], inplace=True)

In [17]:
data.isnull().sum()

Survived    0
Pclass      0
Sex         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [18]:
data.shape

(997760, 7)

In [19]:
data['Fare'] = data['Fare'].astype('int8')

In [20]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked
0,1,1,female,0,0,76,Cherbourg
1,0,3,male,0,0,10,Southampton
2,0,3,male,0,0,12,Cherbourg
3,0,3,male,0,0,13,Southampton
4,0,3,male,0,0,4,Cherbourg


### Label Encoder : 
1. Binary Encoding : 0 and 1
2. Ordinal Encoding : Whenever categories have some order or hierarchy. For example, low, medium, high.
3. nominal Encoding : Whenever categories do not have any order or hierarchy. For example, red, green, blue.



### Feature Encoding :


In [21]:
## Encode = Gender --> Label Encoder

In [22]:
from sklearn.preprocessing import LabelEncoder

In [23]:
le = LabelEncoder()

In [24]:
data['Sex'] = le.fit_transform(data['Sex'])

In [25]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked
0,1,1,0,0,0,76,Cherbourg
1,0,3,1,0,0,10,Southampton
2,0,3,1,0,0,12,Cherbourg
3,0,3,1,0,0,13,Southampton
4,0,3,1,0,0,4,Cherbourg


In [26]:
# data = data.iloc[:,:-1]

In [27]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked
0,1,1,0,0,0,76,Cherbourg
1,0,3,1,0,0,10,Southampton
2,0,3,1,0,0,12,Cherbourg
3,0,3,1,0,0,13,Southampton
4,0,3,1,0,0,4,Cherbourg


In [28]:
from sklearn.preprocessing import OneHotEncoder

In [29]:
onehot = OneHotEncoder(sparse_output=False)
## One hot encoding you don't need sparse output

In [30]:
Embarked = onehot.fit_transform(data[['Embarked']])

In [31]:
Embarked = pd.DataFrame(Embarked, columns=onehot.get_feature_names_out(['Embarked']))

In [32]:
Embarked.head()

,Embarked_Cherbourg,Embarked_Queenstown,Embarked_Southampton
0,1.0,0.0,0.0
1,0.0,0.0,1.0
2,1.0,0.0,0.0
3,0.0,0.0,1.0
4,1.0,0.0,0.0


In [33]:
data = pd.concat([data.drop(columns=['Embarked']), Embarked], axis=1)

In [34]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked_Cherbourg,Embarked_Queenstown,Embarked_Southampton
0,1.0,1.0,0.0,0.0,0.0,76.0,1.0,0.0,0.0
1,0.0,3.0,1.0,0.0,0.0,10.0,0.0,0.0,1.0
2,0.0,3.0,1.0,0.0,0.0,12.0,1.0,0.0,0.0
3,0.0,3.0,1.0,0.0,0.0,13.0,0.0,0.0,1.0
4,0.0,3.0,1.0,0.0,0.0,4.0,1.0,0.0,0.0


### Columns :
mean = 0 --> 
std = 1  ---> scale down the data

### Feature Scaling :
1. Standardization : It is a technique that transforms the data to have a mean of 0 and a standard deviation of 1. It is also known as Z-score normalization. It is useful when the data has a normal distribution and when the features have different scales. It is less sensitive to outliers than Min-Max Scaling.
2. Min-Max Scaling : It is a technique that transforms the data to a fixed range, usually [0, 1]. It is useful when the data does not have a normal distribution and when the features have different scales. It is sensitive to outliers, as it uses the minimum and maximum values of the data to scale it.

In [35]:
from sklearn.preprocessing import StandardScaler

In [36]:
scale = StandardScaler()

In [37]:
data.columns

Index(['Survived', 'Pclass', 'Sex', 'SibSp', 'Parch', 'Fare',
       'Embarked_Cherbourg', 'Embarked_Queenstown', 'Embarked_Southampton'],
      dtype='object')

In [38]:
num_cols = ['Pclass','SibSp', 'Parch', 'Fare']

In [39]:
data[num_cols] = scale.fit_transform(data[num_cols])

In [40]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked_Cherbourg,Embarked_Queenstown,Embarked_Southampton
0,1.0,-1.568865,0.0,-0.462644,-0.469267,1.675922,1.0,0.0,0.0
1,0.0,0.824107,1.0,-0.462644,-0.469267,-0.234102,0.0,0.0,1.0
2,0.0,0.824107,1.0,-0.462644,-0.469267,-0.176222,1.0,0.0,0.0
3,0.0,0.824107,1.0,-0.462644,-0.469267,-0.147283,0.0,0.0,1.0
4,0.0,0.824107,1.0,-0.462644,-0.469267,-0.407741,1.0,0.0,0.0


UI --> Streamlit - website
Scaler, Encoder --> Encoder the data into pickled file and use it in the streamlit app

- 80% --> Train
- 20% --> Test

In [41]:
import pickle

In [42]:
with open('label_encoder.pkl', 'wb') as f:
    pickle.dump(le, f)


In [43]:
with open('onehot_encoder.pkl', 'wb') as f:
    pickle.dump(onehot, f)
    

In [44]:
with open('scaler.pkl', 'wb') as f:
    pickle.dump(scale, f)

## Train - Validation - Test Split

In [45]:
data.head()

,Survived,Pclass,Sex,SibSp,Parch,Fare,Embarked_Cherbourg,Embarked_Queenstown,Embarked_Southampton
0,1.0,-1.568865,0.0,-0.462644,-0.469267,1.675922,1.0,0.0,0.0
1,0.0,0.824107,1.0,-0.462644,-0.469267,-0.234102,0.0,0.0,1.0
2,0.0,0.824107,1.0,-0.462644,-0.469267,-0.176222,1.0,0.0,0.0
3,0.0,0.824107,1.0,-0.462644,-0.469267,-0.147283,0.0,0.0,1.0
4,0.0,0.824107,1.0,-0.462644,-0.469267,-0.407741,1.0,0.0,0.0


In [46]:
data.isnull().sum()

Survived                2234
Pclass                  2234
Sex                     2234
SibSp                   2234
Parch                   2234
Fare                    2234
Embarked_Cherbourg      2234
Embarked_Queenstown     2234
Embarked_Southampton    2234
dtype: int64

In [47]:
data = data.dropna()

In [48]:
data.isnull().sum()

Survived                0
Pclass                  0
Sex                     0
SibSp                   0
Parch                   0
Fare                    0
Embarked_Cherbourg      0
Embarked_Queenstown     0
Embarked_Southampton    0
dtype: int64

In [49]:
X = data.drop(columns=['Survived'])
y = data['Survived']

In [50]:
X.head()

,Pclass,Sex,SibSp,Parch,Fare,Embarked_Cherbourg,Embarked_Queenstown,Embarked_Southampton
0,-1.568865,0.0,-0.462644,-0.469267,1.675922,1.0,0.0,0.0
1,0.824107,1.0,-0.462644,-0.469267,-0.234102,0.0,0.0,1.0
2,0.824107,1.0,-0.462644,-0.469267,-0.176222,1.0,0.0,0.0
3,0.824107,1.0,-0.462644,-0.469267,-0.147283,0.0,0.0,1.0
4,0.824107,1.0,-0.462644,-0.469267,-0.407741,1.0,0.0,0.0


In [51]:
y.head()

0    1.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Survived, dtype: float64

In [52]:
from sklearn.model_selection import train_test_split

In [53]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=20, random_state=42)

In [54]:
X_train,X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

In [55]:
X_train.shape

(796404, 8)

In [56]:
X_valid.shape

(199102, 8)

### Lec 2:  ANN - Artificial Neural Network Implementation

#### Sequential Model  --> Sequence of layers
1. Input Layer --> 2. Hidden Layer --> 3. Output Layer

In [57]:
X_train.shape

(796404, 8)

In [58]:
X_valid.shape

(199102, 8)

If 2 inputs --> 3 neurons in 1st hidden layer --> 3 in 2nd layer --> 1 neuron in output layer
- 2* 3+ 3 + 3*3 + 3 + 3*1 + 1 = 25 parameters --> In forward propagation.

### Sequential Model : ANN Implementation
1. Sequential Network
2. Dense Layer : It is hidden layer in the ANN. In this we mention the number of neurons in the hidden layer and the activation function to be used in that layer. The number of neurons in the hidden layer is a hyperparameter that we need to tune. The activation function is also a hyperparameter that we need to tune.
3. Activation Function : It is a function that is applied to the output of each neuron in the hidden layer. It is used to introduce non-linearity in the model. It helps the model to learn complex patterns in the data. Some common activation functions are ReLU, Sigmoid, Tanh, etc. The activation function should be sigmoid for the output layer because it is a binary classification problem. If it is a multi-class classification problem, we will use softmax activation function in the output layer. If it is a regression problem, we will not use any activation function in the output layer.
4. Optimizer : It is an algorithm that is used to update the weights of the neurons in the hidden layer during the training process. It is used to minimize the loss function. Some common optimizers are SGD, AdaGrad,Adadelta, Adam, RMSprop, etc.
5. Loss function : It is a function that is used to measure the error between the predicted output and the actual output. It is used to train the model. Some common loss functions are Mean Squared Error, Mean Absolute Error, Binary Crossentropy, Categorical Crossentropy, etc. It is a classification problem, we will use Binary Crossentropy loss function. If it is a multi-class classification problem, we will use Categorical Crossentropy loss function. If it is a regression problem, we will use Mean Squared Error loss function.
6. Metrics : It is a function that is used to evaluate the performance of the model. Some common metrics are Accuracy, Precision, Recall, F1 Score, etc. It is a classification problem, we will use Accuracy as the metric. If it is a multi-class classification problem, we will use Accuracy as the metric. If it is a regression problem, we will use Mean Absolute Error as the metric.
7. TensorBoard : It is a tool that is used to visualize the training process of the model. It is used to monitor the loss and accuracy of the model during the training process. It is also used to visualize the weights of the neurons in the hidden layer.
4. Output Layer : It is the last layer in the ANN. It is used to make predictions. The number of neurons in the output layer depends on the type of problem we are trying to solve. For example, if we are trying to solve a binary classification problem, we will have 1 neuron in the output layer. If we are trying to solve a multi-class classification problem, we will have as many neurons in the output layer as the number of classes. If we are trying to solve a regression problem, we will have 1 neuron in the output layer.

In [59]:
import tensorflow as tf

2026-04-17 17:55:21.806542: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2026-04-17 17:55:21.812893: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1776428721.820726   14646 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1776428721.823228   14646 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2026-04-17 17:55:21.831125: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instr

### Model Building :

In [60]:
import tensorflow as tf
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.callbacks import EarlyStopping, TensorBoard

print('TensorFlow:', tf.__version__)

TensorFlow: 2.18.0


In [61]:
(X_train.shape[1],)
## In sequential Layer always give the input shape of the 
# data in the first layer only. 
# After that, it will automatically 
# understand the shape of the data.
## As list as given above.

(8,)

In [62]:
model = Sequential([Dense(64, activation='relu', input_shape=(X_train.shape[1],)), ## First layer with input shape
            Dense(32, activation='relu'), ## Second layer
            Dense(1, activation='sigmoid')]) ## Output layer with sigmoid activation for binary classification


/home/dharmender/anaconda3/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
I0000 00:00:1776428722.957156   14646 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 1216 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3050 6GB Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.6


In [63]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 64)             │           576 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,689 (10.50 KB)

 Trainable params: 2,689 (10.50 KB)

 Non-trainable params: 0 (0.00 B)

Parameters :
- First layer : 64 neurons --> 64 * (number of input features) + 64 (bias) = 64 * (number of input features) + 64 = 64 * x  +  64  = 576
- Second layer : 32 neurons --> 32 * 64 + 32 (bias) = 32 * 64 + 32 = 2080
- Output layer : 1 neuron --> 1 * 32 + 1 (bias) = 33


### Model Compilation : --> Optimizer, Loss function, Metrics 
It is the process of configuring the model for training. It is done before training the model. It is used to specify the optimizer, loss function, and metrics to be used during the training process. The optimizer is used to update the weights of the neurons in the hidden layer during the training process. The loss function is used to measure the error between the predicted output and the actual output. The metrics are used to evaluate the performance of the model.

In [64]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [65]:
### adam --> learning rate = 0.001 (default) --> It is the step size at which the model learns. It is used to update the weights of the neurons in the hidden layer during the training process. 
# The learning rate is a hyperparameter that can be tuned to improve the performance of the model.
# A higher learning rate can lead to faster 
# convergence but may also lead to overshooting 
# the optimal weights, while a lower learning rate 
# can lead to slower convergence but may also lead to 
# better performance.

If want to create the optimizer by your own then we can use the following code :

In [66]:
opt = keras.optimizers.Adam(learning_rate=0.01)

In [67]:
model.compile(optimizer=opt, loss='binary_crossentropy', metrics=['accuracy'])

In [68]:
stopping_callback = EarlyStopping(monitor='val_loss', patience=5, restore_best_weights=True)


In [69]:
model.fit(X_train,y_train, validation_data=(X_valid, y_valid), epochs=50, callbacks=[stopping_callback])

Epoch 1/50


I0000 00:00:1776428724.019287   15675 service.cc:148] XLA service 0x73fc140034d0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1776428724.019306   15675 service.cc:156]   StreamExecutor device (0): NVIDIA GeForce RTX 3050 6GB Laptop GPU, Compute Capability 8.6
2026-04-17 17:55:24.031606: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:268] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
I0000 00:00:1776428724.101400   15675 cuda_dnn.cc:529] Loaded cuDNN version 91700


  239/24888 ━━━━━━━━━━━━━━━━━━━━ 16s 665us/step - accuracy: 0.7868 - loss: 0.4664

I0000 00:00:1776428724.678937   15675 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


24888/24888 ━━━━━━━━━━━━━━━━━━━━ 19s 730us/step - accuracy: 0.8375 - loss: 0.3637 - val_accuracy: 0.8525 - val_loss: 0.3335
Epoch 2/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 18s 722us/step - accuracy: 0.8504 - loss: 0.3378 - val_accuracy: 0.8536 - val_loss: 0.3315
Epoch 3/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 17s 701us/step - accuracy: 0.8493 - loss: 0.3362 - val_accuracy: 0.8527 - val_loss: 0.3335
Epoch 4/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 17s 690us/step - accuracy: 0.8509 - loss: 0.3342 - val_accuracy: 0.8543 - val_loss: 0.3294
Epoch 5/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 19s 773us/step - accuracy: 0.8512 - loss: 0.3342 - val_accuracy: 0.8527 - val_loss: 0.3345
Epoch 6/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 19s 773us/step - accuracy: 0.8502 - loss: 0.3354 - val_accuracy: 0.8520 - val_loss: 0.3316
Epoch 7/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 19s 771us/step - accuracy: 0.8509 - loss: 0.3343 - val_accuracy: 0.8534 - val_loss: 0.3334
Epoch 8/50
24888/24888 ━━━━━━━━━━━━━━━━━━━━ 19s 757us/step - accur

### Always save in the h5 format 
Because it is the best format for saving the model. It is a binary format that is used to save the model architecture, weights, and optimizer state. It is also used to save the model in a compressed format. It is also used to save the model in a format that can be easily loaded and used for inference.

In [70]:
model.save('titanic_model.keras')  # recommended format

### Lectures 3 : Model Prediction using ANN
- scaler -> pickled file
- encoder -> pickled file
- model -> h5 file
